Feature/join-room-token

escalopa opened the pull request.

lordvidex committed:
feat(env): changed package to viper

- reason is because `.env` file was not read by the other packages that
  I found.
- tradeoff is the size of the `viper` package

Commit:
feat: join room by token

Commit:
feat: save game on close

Commit:
feat: create a resp type for game

lordvidex commented on handler/handler.go:
```diff
@@ -194,13 +233,55 @@ func (h *Handler) room(w http.ResponseWriter, r *http.Request) {
+	}
+}
+
+func toGuess(s game.Session) []guessResponse {
+	sr := make([]guessResponse, len(s.Guesses))
+	var status []int
+	for _, g := range s.Guesses {
+		for _, v := range g.Stats {
+			status = append(status, int(v))
+		}
+		sr = append(sr, guessResponse{
+			Word:     g.Word,
+			PlayedAt: g.PlayedAt.Time,
+			Status:   status,
+		})
+	}
+	return sr
+}
```
the `status` here is not correct. the second player will inherit the `status` of the first user because you are not recreating the slice.

lordvidex commented on handler/live.go:
```diff
@@ -23,42 +23,32 @@ var (
 	// Get room from Hub
 	Hub.mu.Lock()
-	defer Hub.mu.Unlock()
-	room, ok := Hub.rooms[roomID]
+	room, ok := Hub.rooms[gameID]
+	Hub.mu.Unlock()
 	if !ok {
```
RWMutex will be better to use for the Hub because most of the times, when we need a room, we have a lot of reads, and it doesn't make sense for us to lock other readers too.

lordvidex commented on service/random/random.go:
```diff
@@ -0,0 +1,76 @@
+	s map[string]value
+}
+
+// New returns a new RandomGen
+func New() RandomGen {
+	r := RandomGen{
+		r: rand.New(rand.NewSource(time.Now().UnixNano())),
+		l: defaultLenght,
+		s: make(map[string]value),
+	}
+	go r.cleanup()
+	return r
+}
```
RandomGen should listen to some context, to end cleanup when the server is shutting down for efficient resource cleanup.

lordvidex commented on service/random/random.go:
```diff
@@ -0,0 +1,76 @@
+
+// Store stores the username and gameID associated with a the token and returns it
+func (rg RandomGen) Store(username string, gameID uuid.UUID) string {
+	var token string
+	for i := 0; i < rg.l; i++ {
+		token += string(characters[rg.r.Intn(len(characters))])
+	}
```
`string.Builder` or slice of runes >> string appending

lordvidex committed:
chore: little fixes

lordvidex committed:
fix: fixed the guessResponse struct in handler

link: https://github.com/Chat-Map/wordle-server/pull/7#discussion_r1232387638

lordvidex committed:
fix: changed Mutex to RWMutex for GameHub

link: https://github.com/Chat-Map/wordle-server/pull/7#discussion_r1232391215

lordvidex committed:
fix: updated random gen

- added appCtx to RandomGenerator clean func
- removed string concat

link: https://github.com/Chat-Map/wordle-server/pull/7#discussion_r1232396429
link: https://github.com/Chat-Map/wordle-server/pull/7#discussion_r1232395200

lordvidex commented:
https://github.com/Chat-Map/wordle-server/issues/8

lordvidex committed:
fix(join-room): prevented crash when usersession is nil

Commit:
added leaderboard
